cpp/src/arrow/buffer.cc

```diff
@@ -155,10 +155,23 @@ Status AllocateBuffer(MemoryPool* pool, const int64_t size,
   return Status::OK();
 }
 
+Status AllocateBuffer(MemoryPool* pool, const int64_t size,
+                      std::unique_ptr<Buffer>* out) {
+  auto buffer = std::unique_ptr<PoolBuffer>(new PoolBuffer(pool));
+  RETURN_NOT_OK(buffer->Resize(size));
+  buffer->ZeroPadding();
+  *out = std::move(buffer);
+  return Status::OK();
```

Comment:
This could be templated something like

```c++
template <typename BufferType, typename Container>
inline Status ReturnBufferSized(Container<BufferType>&& buffer, const int64_t size,
                                Container<Buffer>* out) {
  RETURN_NOT_OK(buffer->Resize(size));
  buffer->ZeroPadding();
  *out = std::move(buffer);
  return Status::OK();
}
```
This can also be used to address code duplication in `AllocateResizableBuffer` as it is now